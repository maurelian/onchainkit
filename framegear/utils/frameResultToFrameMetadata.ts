import { FrameMetadataType } from '@coinbase/onchainkit';

export function frameResultToFrameMetadata(result: Record<string, string>): FrameMetadataType {
  const buttons = [1, 2, 3, 4].map((idx) =>
    result[`fc:frame:button:${idx}`]
      ? {
          action: result[`fc:frame:button:${idx}:action`],
          label: result[`fc:frame:button:${idx}`],
          target: result[`fc:frame:button:${idx}:target`],
        }
      : undefined,
  );
  const image = result['fc:frame:image'];
  const inputText = result['fc:frame:input'];
  const input = inputText ? { text: inputText } : undefined;
  const postUrl = result['fc:frame:post_url'];
  const rawState = result['fc:frame:state'];
  const state = rawState ? JSON.parse(result['fc:frame:state']) : undefined;

  return { buttons: buttons as any, image, input, postUrl, state };
}
