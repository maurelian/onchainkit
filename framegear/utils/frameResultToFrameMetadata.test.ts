import { frameResultToFrameMetadata } from './frameResultToFrameMetadata';

const baseGoodDefinition = {
  'fc:frame': 'vNext',
  'fc:frame:image': 'https://images.party/this-is-a-test.apng',
  'og:image': 'https://images.party/this-is-a-test.apng',
};

describe('frameResultToFrameMetadata', () => {
  it('handles the minimum viable input', () => {
    expect(frameResultToFrameMetadata(baseGoodDefinition)).toEqual(
      expect.objectContaining({
        image: baseGoodDefinition['og:image'],
      }),
    );
  });

  it('handles buttons', () => {
    expect(
      frameResultToFrameMetadata({
        ...baseGoodDefinition,
        'fc:frame:button:1': 'henlo',
        'fc:frame:button:1:action': 'post',
        'fc:frame:button:2': 'henlo',
        'fc:frame:button:2:action': 'post_redirect',
        'fc:frame:button:3': 'henlo',
        'fc:frame:button:3:action': 'mint',
        'fc:frame:button:4': 'henlo',
        'fc:frame:button:4:action': 'link',
        'fc:frame:button:4:target': 'https://google.com',
      }),
    ).toMatchSnapshot();
  });

  it('handles input text', () => {
    expect(frameResultToFrameMetadata({ ...baseGoodDefinition, 'fc:frame:input': 'lmbo' })).toEqual(
      expect.objectContaining({
        input: { text: 'lmbo' },
      }),
    );
  });

  it('handles state', () => {
    expect(
      frameResultToFrameMetadata({
        ...baseGoodDefinition,
        'fc:frame:state': JSON.stringify({ status: 'buzzin' }),
      }),
    ).toEqual(
      expect.objectContaining({
        state: { status: 'buzzin' },
      }),
    );
  });
});
