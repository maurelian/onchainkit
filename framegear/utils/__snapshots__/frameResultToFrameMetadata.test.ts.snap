// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frameResultToFrameMetadata handles buttons 1`] = `
{
  "buttons": [
    {
      "action": "post",
      "label": "henlo",
      "target": undefined,
    },
    {
      "action": "post_url",
      "label": "henlo",
      "target": undefined,
    },
    {
      "action": "mint",
      "label": "henlo",
      "target": undefined,
    },
    {
      "action": "link",
      "label": "henlo",
      "target": "https://google.com",
    },
  ],
  "image": "https://images.party/this-is-a-test.apng",
  "input": undefined,
  "postUrl": undefined,
  "state": undefined,
}
`;
